src/uu/lsmem/src/lsmem.rs

```diff
@@ -3,15 +3,543 @@
+//     },
+//     ColDesc {
+//         name: "ZONES",
+//         whint: 0.0,
+//         flags: SCOLS_FL_RIGHT,
+//         help: "valid zones for the memory range",
+//     },
+// ];
+
+#[derive(Debug, Deserialize, PartialEq, Clone, Copy)]
+enum ZoneId {
+    #[serde(rename = "ZONE_DMA")]
+    ZoneDma,
+    #[serde(rename = "ZONE_DMA32")]
+    ZoneDma32,
+    #[serde(rename = "ZONE_NORMAL")]
+    ZoneNormal,
+    #[serde(rename = "ZONE_HIGHMEM")]
+    ZoneHighmem,
+    #[serde(rename = "ZONE_MOVABLE")]
+    ZoneMovable,
+    #[serde(rename = "ZONE_DEVICE")]
+    ZoneDevice,
+    #[serde(rename = "ZONE_NONE")]
+    ZoneNone,
+    #[serde(rename = "ZONE_UNKNOWN")]
+    ZoneUnknown,
+    #[serde(rename = "MAX_NR_ZONES")]
+    MaxNrZones,
+}
+
+#[derive(PartialEq, Clone)]
+enum MemoryState {
+    Online,
+    Offline,
+    GoingOffline,
+    Unknown,
+}
+
+impl core::fmt::Display for MemoryState {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        match self {
+            MemoryState::Online => write!(f, "online"),
+            MemoryState::Offline => write!(f, "offline"),
+            MemoryState::GoingOffline => write!(f, "going-offline"),
+            MemoryState::Unknown => write!(f, "unknown"),
+        }
+    }
+}
+
+impl FromStr for MemoryState {
+    type Err = ();
+    fn from_str(input: &str) -> Result<MemoryState, Self::Err> {
+        match input {
+            "online" => Ok(MemoryState::Online),
+            "offline" => Ok(MemoryState::Offline),
+            "going-offline" => Ok(MemoryState::GoingOffline),
+            "unknown" => Ok(MemoryState::Unknown),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Clone)]
+struct MemoryBlock {
+    index: u64,
+    count: u64,
+    state: MemoryState,
+    node: i32,
+    nr_zones: usize,
+    zones: [ZoneId; ZoneId::MaxNrZones as usize],
+    removable: bool,
+}
+
+impl MemoryBlock {
+    fn new() -> Self {
+        MemoryBlock {
+            index: 0,
+            count: 0,
+            state: MemoryState::Unknown,
+            node: 0,
+            nr_zones: 0,
+            zones: [ZoneId::ZoneUnknown; ZoneId::MaxNrZones as usize],
+            removable: true,
+        }
+    }
+}
+
+#[derive(Tabled)]
+struct TableRow {
+    #[tabled(rename = "RANGE")]
+    range: String,
+    #[tabled(rename = "SIZE")]
+    size: String,
+    #[tabled(rename = "STATE")]
+    state: String,
+    #[tabled(rename = "REMOVABLE")]
+    removable: String,
+    #[tabled(rename = "BLOCK")]
+    block: String,
+    #[tabled(rename = "NODE")]
+    node: String,
+    #[tabled(rename = "ZONES")]
+    zones: String,
+}
+
+impl TableRow {
+    fn new() -> Self {
+        TableRow {
+            range: String::new(),
+            size: String::new(),
+            state: String::new(),
+            removable: String::new(),
+            block: String::new(),
+            node: String::new(),
+            zones: String::new(),
+        }
+    }
+}
```

Comment:
I think this `impl` is not necessary. Instead, you could derive from `Default` and then use `default()` instead of `new()`.

Reply:
removed